changes/2020-06-16_required-examples/change.md

```diff
@@ -0,0 +1,98 @@
+[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
+[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"
+
+# Define Required Examples
+
+## Affected Specifications
+
+This serves as a reference of all specification documents that this change affects.
+
+| Specification                                         |
+| ----------------------------------------------------- |
+| [examples](../../examples/examples.md)                |
+| [examples readme](../../examples/templates/readme.md) |
+
+## Affected Implementations
+
+This serves as a reference for all implementations that this change affects.
+
+| Language   | Repository                                                                            |
+| ---------- | ------------------------------------------------------------------------------------- |
+| Python     | [aws-encryption-sdk-c](https://github.com/aws/aws-encryption-sdk-python)              |
+| Java       | [aws-encryption-sdk-javascript](https://github.com/aws/aws-encryption-sdk-java)       |
+| C          | [aws-encryption-sdk-c](https://github.com/aws/aws-encryption-sdk-c)                   |
+| Javascript | [aws-encryption-sdk-javascript](https://github.com/aws/aws-encryption-sdk-javascript) |
+
+## Definitions
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Summary
+
+This document defines the examples that each implementation MUST include.
+
+## Out of Scope
+
+The specific implementations details for each example
+are out of scope for this document.
+The template document for each example contains those details.
+
+## Motivation
+
+The two primary motivations for requiring examples are
+to provide guidance to AWS Encryption SDK customers
+and to force AWS Encryption SDK developers
+to use the APIs that they design and build
+to verify that they provide the intended user experience.
+
+## Drawbacks
+
+Requiring each implementation to include many examples
+increases the level of effort to create each implementation.
+We believe that the value to customers of
+improving the usability of APIs
+and providing guidance on how to use them
+justifies this cost.
+
+## Security Implications
+
+We MUST assume that any example will be used in production.
```

Comment:
Since these example MUST run under CI,
we should be explicit that all keys
come from the test vectors.

Reply:
I'm hesitant to blend examples and test vectors, mainly because it feels like it is over-complicating matters. I could see the argument from the perspective of making it even more clear that you need to do *something* to manage your wrapping keys in production, but at the same time I worry that it might go the other direction and come across as "here is an example of *how* you might manage your wrapping keys", which is explicitly not what we are trying to show.

Reply:
Ok, then what is at stake for me, is minimizing the number of wrapping keys that are floating around. Since we have to assume that this code will be copied and used in production, I want to minimize peoples ability to accidentally copy an RSA private key...

Reply:
This is one reason why for the current example implementations we generate a new wrapping key inside the example[1]. Hopefully we have made it sufficiently clear in the comments that you need to get your wrapping keys from somewhere else in production, but worst case if someone does copy-paste into production, the failure mode is data loss (because the wrapping key is never stored) rather than data compromise (because a publicly accessible wrapping key is used). Neither of these things is *good* but data loss is *less bad*.

[1] https://github.com/aws/aws-encryption-sdk-python/blob/5628ede2564f1b2003088d7f3901cda87ce0a0de/examples/src/keyring/raw_rsa/keypair.py#L38-L46